src/Eval.hs

```diff
@@ -848,20 +848,65 @@ specialCommandSet ctx [x@(XObj (Sym path@(SymPath mod n) _) _ _), value] = do
+                -- At the moment, let bindings are not structured the same as global defs or dynamic defs.
+                -- This makes calls to the type check problematic, as we cannot work against a common binding form.
+                -- Once we better support let bindings, type check them.
+                -- (ctx', typedVal) <- typeCheckValueAgainstBinder ctx value binder
+                return $ if contextPath ctx == mod
+                         then either (failure ctx) (success ctx) (Right value)
+                         else (ctx, dynamicNil)
+                where success c xo = (c{contextInternalEnv = Just (setStaticOrDynamicVar (SymPath [] n) env binder xo)}, dynamicNil)
+              -- If the def isn't found in the internal environment, check the global environment.
+              Nothing -> setGlobal ctx (contextGlobalEnv ctx) value
 specialCommandSet ctx [notName, body] =
   return (evalError ctx ("`set!` expected a name as first argument, but got " ++ pretty notName) (info notName))
 specialCommandSet ctx args =
   return (evalError ctx ("`set!` takes a name and a value, but got `" ++ intercalate " " (map pretty args)) (if null args then Nothing else info (head args)))
 
+-- | Convenience method for signifying failure in a given context.
+failure :: Context -> EvalError -> (Context, Either EvalError a)
+failure ctx err = (ctx, Left err)
+
+-- | Given a context, value XObj and an existing binder, check whether or not
+-- the given value has a type matching the binder's in the given context.
+typeCheckValueAgainstBinder :: Context -> XObj -> Binder -> IO (Context, (Either EvalError XObj))
+typeCheckValueAgainstBinder ctx val binder = do
+  (ctx', typedValue) <- annotateWithinContext False ctx val
+  case typedValue of
+    Right (val', deps) -> return (go ctx' binderTy val')
+    Left err -> return (ctx', Left err)
+  where path = (getPath (binderXObj binder))
+        binderTy = ty (binderXObj binder)
+        typeErr x = evalError ctx ("can't `set!` " ++ show path ++ " to a value of type " ++ show (fromJust (ty x)) ++ ", " ++ show path ++ " has type " ++ show (fromJust binderTy)) (info x)
+        go ctx (Just DynamicTy) x = (ctx, Right x)
+        go ctx t x@(XObj _ _ t') = if t == t' then (ctx, Right x) else typeErr x
+
+-- | Sets a variable, checking whether or not it is static or dynamic, and
+-- assigns an appropriate type to the variable.
+-- Returns a new environment containing the assignment.
+setStaticOrDynamicVar :: SymPath -> Env -> Binder -> XObj -> Env
+setStaticOrDynamicVar path env binder value =
+    case binder of
+      (Binder meta (XObj (Lst (def@(XObj Def _ _) : sym : val)) i t)) ->
+        envReplaceBinding path (Binder meta (XObj (Lst [def, sym, value]) (info value) t)) env
+      (Binder meta (XObj (Lst (defdy@(XObj DefDynamic _ _) : sym : val)) i t)) ->
+        envReplaceBinding path (Binder meta (XObj (Lst [defdy, sym, value]) (info value) (Just DynamicTy))) env
+      _ -> env
```

Comment:
When does this happen? Would an error be more appropriate here?

Reply:
Ah, good call!

Reply:
I've added a `TODO` for now--probably what we'll want to do is just return an either here and let callers decide what to do with the result when it's an error